cpp/src/arrow/stl_iterator.h

```diff
@@ -128,6 +131,199 @@ class ArrayIterator {
+  }
+  ChunkedArrayIterator& operator+=(difference_type n) {
+    bool in_current_chunk;
+    auto& current_iterator = iterators_list_[current_chunk_index_];
+    if (n >= 0) {
+      in_current_chunk =
+          n < chunked_array_->chunk(static_cast<int>(current_chunk_index_))->length() -
+                  current_iterator.index();
+    } else {
+      in_current_chunk = n > current_iterator.index();
+    }
+    if (in_current_chunk) {
+      index_ += n;
+      current_iterator += n;
+      return *this;
+    } else {
+      current_iterator -= current_iterator.index();
+      index_ += n;
+      UpdateComponents(*this, true);
+      return *this;
+    }
+  }
+  ChunkedArrayIterator& operator-=(difference_type n) {
+    (*this) += -n;
+    return *this;
+  }
+
+  // Comparisons
+  bool operator==(const ChunkedArrayIterator& other) const {
+    return index_ == other.index_;
+  }
+  bool operator!=(const ChunkedArrayIterator& other) const {
+    return index_ != other.index_;
+  }
+  bool operator<(const ChunkedArrayIterator& other) const {
+    return index_ < other.index_;
+  }
+  bool operator>(const ChunkedArrayIterator& other) const {
+    return index_ > other.index_;
+  }
+  bool operator<=(const ChunkedArrayIterator& other) const {
+    return index_ <= other.index_;
+  }
+  bool operator>=(const ChunkedArrayIterator& other) const {
+    return index_ >= other.index_;
+  }
+
+ private:
+  arrow::internal::ChunkLocation GetChunkLocation(int64_t index) const {
+    return chunked_array_->GetChunkResolver().Resolve(index);
+  }
+
+  void UpdateComponents(ChunkedArrayIterator<ArrayType>& chunked_array_iterator,
+                        bool update = false) {
+    if (!update) {
+      int64_t chunk_index = 0;
+      for (const auto& array : chunked_array_iterator.chunked_array_->chunks()) {
+        chunked_array_iterator.iterators_list_.emplace_back(
+            *arrow::internal::checked_pointer_cast<ArrayType>(array));
+        auto chunk_length = array->length();
+        if (chunk_index) {
+          chunk_length += chunked_array_iterator.chunks_lengths_[chunk_index - 1];
+        }
+        chunked_array_iterator.chunks_lengths_.push_back(chunk_length);
+        chunk_index++;
+      }
+    }
+
+    chunked_array_iterator.current_chunk_index_ =
+        GetChunkLocation(chunked_array_iterator.index_).chunk_index;
+    auto& current_iterator =
+        chunked_array_iterator
+            .iterators_list_[chunked_array_iterator.current_chunk_index_];
+    current_iterator -= current_iterator.index() - chunked_array_iterator.index_;
+    if (chunked_array_iterator.current_chunk_index_) {
+      current_iterator -=
+          chunked_array_iterator
+              .chunks_lengths_[chunked_array_iterator.current_chunk_index_ - 1];
+    }
+  }
+
+  const ChunkedArray* chunked_array_;
+  int64_t index_;
+  int64_t current_chunk_index_;
+  std::vector<int64_t> chunks_lengths_;
+  std::vector<ArrayIterator<ArrayType>> iterators_list_;
```

Comment:
I'm a bit surprised, are these members still needed? Ideally, everything can be done using the ChunkResolver...